### PR TITLE
Containerize e2e tests

### PR DESCRIPTION
**Pull request by rylew1**
## Ticket

https://github.com/navapbc/template-infra/issues/735

## Changes

- containerize e2e tests with playwright image

## Context
- Imrproves portability and avoids having to `apt-get` a bunch of packages https://github.com/navapbc/template-infra/blob/32ce571bd5656c4efc0615c58b2e8ed2a3a739f8/Makefile#L235
- For all local e2e testing - you'll need to have the app running locally and call the run command with the correct port and app (folder name) (hitting your locally running app from local docker uses a base url of `host.docker.internal`)

- PR checks are still testing against the PR preview env 

Opened on a `template-infra` => https://github.com/navapbc/template-infra/pull/770

https://playwright.dev/docs/docker

https://github.com/user-attachments/assets/be6ff595-6a81-49e4-823a-32a978a236b5

## Preview environment
- Service endpoint: http://p-88-app-dev-2033577064.us-east-1.elb.amazonaws.com
- Deployed commit: afdde0ddee55dc7bc75016aa5514e81ca6237b82

### CODE REVIEW

**Comment by rylew1** [comment]
- This works in CI
- Works locally if you use `host.docker.internal` - (just trying to use `Dockerfile` and not a Docker compose to port map things)



Local command if server running on port 3001:
`docker run --name playwright-e2e-container -w /app playwright-e2e make e2e-test APP_NAME=app BASE_URL=http://host.docker.internal:3001`

File: Makefile
Comment (rylew1) [inline Makefile:239]: TODO: these make commands could probably be updated

File: .github/workflows/e2e-tests.yml
Comment (lorenyu) [inline .github/workflows/e2e-tests.yml:37]: what's the purpose of removing the docker container

File: .github/workflows/e2e-tests.yml
Comment (lorenyu) [inline .github/workflows/e2e-tests.yml:24]: we can wrap the docker build  in a make target like `make e2e-build` and `make e2e-run`

File: e2e/Dockerfile
Comment (lorenyu) [inline e2e/Dockerfile:1]: nice this dockerfile looks pretty clean

File: .github/workflows/e2e-tests.yml
Comment (lorenyu) [inline .github/workflows/e2e-tests.yml:29]: are you planning on updating `make e2e-test` to use this command? then we can keep this as `make e2e-test`

**Review by lorenyu — COMMENTED** [review]
Yeah general direction looks good!

**Comment by rylew1** [comment]
I want to test some more but I think this is another review point - posted the docs table for how things are run.

File: e2e/Dockerfile
Comment (lorenyu) [inline e2e/Dockerfile:12]: do we need the app in this dockerfile? i thought this dockerfile was just for playwright. is there a way for playwright to be able to access an app running locally outside of the playwright container? if so that'd be ideal so we can use the single playwright container to test multiple different apps without needing to copy all the apps into the playwright container

**Review by lorenyu — COMMENTED** [review]
looks good, just one main architectural point that it'd be ideal not to have the app inside the e2e container

File: e2e/Dockerfile
Comment (lorenyu) [inline e2e/Dockerfile:22]: nit: the /app folder is a little confusing since it doesn't map to the concept of applications in the rest of the project, so to simplify let's just put the Makefile in the root folder and the e2e folder in /e2e e.g.

```suggestion
WORKDIR /

# Copy the required files and directories from the root of the project into the Docker container
COPY Makefile /Makefile
COPY e2e /e2e

# Install the dependencies inside the /e2e directory
WORKDIR /e2e
RUN npm ci

# Install Playwright browsers and other necessary system dependencies
RUN npx playwright install --with-deps

# Change the working directory back to the root to run the Makefile
WORKDIR /
```

**Review by lorenyu — APPROVED** [review]
Looks good just some nits, and also before merging make sure to test the local commands. I wasn't sure if you retested running locally in Docker after removing the app from the Docker container, but would like to make sure that part works

**Comment by rylew1** [comment]
It looks like most of the comments were nits but still tried to address most of them. Was planning to move changes to `template-infra` now.

>I wasn't sure if you retested running locally in Docker after removing the app from the Docker container, but would like to make sure that part works

I tested natively and in Docker based on the instructions we have in the table and it works

File: Makefile
Comment (doshitan) [inline Makefile:259]: Rather than separately copying the report from the container file system, could consider mounting a location from the host system in the `docker run` call for playwright to write its report to then can update to use `docker run --rm` to keep the containers cleaned up from each run (rather than having to separately call `docker rm` up above as well).

**Review by doshitan — COMMENTED** [review]
Caveat I haven't looked at the actual e2e stuff; just a few thoughts as I was scrolling through this PR, wouldn't call any of them blocking.

File: .github/workflows/e2e-tests.yml
Comment (lorenyu) [inline .github/workflows/e2e-tests.yml:16]: in [pr-environment-checks.yml](https://github.com/navapbc/template-infra/blob/rylew/e2e-containerize/.github/workflows/pr-environment-checks.yml#L68) we should set the name to " " from "Run E2E tests" so it will look cleaner in the PR i.e. it could be "CI App PR Environment Checks / / / E2E Tests (pull_request)" instead of "CI App PR Environment Checks / / Run E2E Tests / E2E Tests (pull_request)"

image

File: docs/e2e/e2e-checks.md
Comment (lorenyu) [inline docs/e2e/e2e-checks.md:33]: extra newlines

```suggestion
```

File: docs/e2e/e2e-checks.md
Comment (lorenyu) [inline docs/e2e/e2e-checks.md:80]: This table is still hard to read in the code, the intent of markdown is that it should be human readable. I don't think a table is needed here. Just write this in prose. Something like

```suggestion
## Run end-to-end tests

First, make sure the application you want to test is running.

Then, run end-to-end tests using Docker with

``
make e2e-test
``

## Run end-to-end tests natively

First, make sure the application you want to test is running.

To run end-to-end tests natively, first install Playwright with

``bash
make e2e-setup-native
``

Then run the tests with

``bash
make e2e-test-native
``
```

**Review by lorenyu — COMMENTED** [review]
Added just a couple more nits. The suggestions from Tanner also seem good to do.

**Comment by rylew1** [comment]
I think we're close to being able to move this to `template-infra` for merging - let me know if there's anything else

File: Makefile
Comment (lorenyu) [inline Makefile:248]: I left this comment in my previous review. It's confusing to have `e2e-run` and `e2e-test`. What's the difference? Why does `run` mean run in Docker and `test` mean run natively? Please rename to `e2e-test` and `e2e-test-native`

File: docs/e2e/e2e-checks.md
Comment (lorenyu) [inline docs/e2e/e2e-checks.md:30]: Style nits:
- Use imperative verbs for instructions "Run tests" rather than "Running tests"
- Use sentence case "Run tests" rather than "Run Tests"

**Review by lorenyu — APPROVED** [review]
some nits otherwise LGTM!

**Pull request closed without merging**
